gh-90622: Do not spawn ProcessPool workers on demand via fork method.

gpshead opened the pull request:
This avoids potential deadlocks in the child processes due to forking from
a multithreaded process.

This is the bugfix only PR suitable for backporting (manual intervention likely still required). See #91587 for the PR fixing the new 3.11 feature that builds on this.  I expect to merge that one first and rebase this on top of it.

gpshead committed:
gh-90622: Do not spawn ProcessPool workers on demand when they spawn via fork.

This avoids potential deadlocks in the child processes due to forking from
a multithreaded process.

gpshead committed:
Merge branch 'main' into disallow-dynamic-process-spawn-with-fork-bugfix-only

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -713,15 +719,27 @@ def _adjust_process_count(self):
-                      self._max_tasks_per_child))
-            p.start()
-            self._processes[p.pid] = p
+            assert self._safe_to_dynamically_spawn_children or not self._executor_manager_thread, 'https://github.com/python/cpython/issues/90622'
```
Hmm, I'm worried this assert may hit and break existing code, can we remove it?

gpshead replied:
I'll definitely remove it entirely from any backport.  I'd _like_ to see if it trips anything up in 3.11beta1 before deciding to remove it though. It _should_ only come up if people are calling this non-public API themselves (subclassing?). Perhaps turning it into a warning for 3.11 is safer?

pitrou replied:
`_adjust_process_count` can also be called if a worker process exited (for example because of a crash).
We want this case to continue working even on a "fork" context.

gpshead replied:
assertion disabled. comment added. we still have a bug here, though it should be a rare one: if a 'fork' mp_context worker dies, we'll spawn (fork) a replacement despite knowing that a thread is running.

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -713,15 +719,27 @@ def _adjust_process_count(self):
+            assert self._safe_to_dynamically_spawn_children or not self._executor_manager_thread, 'https://github.com/python/cpython/issues/90622'
+            self.__spawn_process()
+
+    def __launch_processes(self):
```
The use of double leading underscores here seems gratuitous, can we stick to the existing of a single leading underscore for internal methods?

gpshead committed:
Use single leading _s, disable assertion.

Adds a comment describing why the lack of assertion means we still have
a potential bug.

gpshead committed:
Merge branch 'main' into disallow-dynamic-process-spawn-with-fork-bugfix-only

gpshead committed:
Merge branch 'main' into disallow-dynamic-process-spawn-with-fork-bugfix-only

gpshead committed:
Skip test_idle_process tests on "fork"

gpshead committed:
Merge branch 'main' into disallow-dynamic-process-spawn-with-fork-bugfix-only

gpshead committed:
Merge branch 'main' into disallow-dynamic-process-spawn-with-fork-bugfix-only
